### PR TITLE
lib/shapelib: upgrade from upstream

### PR DESCRIPTION
**Pull request by nilason**
Upgrade with upstream files from GDAL 3.5.3 and SHAPELIB 1.5.0 .

### CODE REVIEW

**Comment by nilason** [comment]
Initially I aimed to fix some compiler warnings (from Ubuntu CI builds):

```
gcc  -g -O2  -fPIC  -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include    -DPACKAGE=\""grasslibs"\"   -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -I/home/runner/work/grass/grass/dist.x86_64-pc-linux-gnu/include -DRELDIR=\"lib/external/shapelib\" -o OBJ.x86_64-pc-linux-gnu/dbfopen.o -c dbfopen.c

...

In file included from /usr/include/string.h:495,
                 from dbfopen.c:166:
In function ‘strncpy’,
    inlined from ‘DBFAddNativeFieldType’ at dbfopen.c:908:9:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: warning: ‘__builtin_strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dbfopen.c: In function ‘DBFAddNativeFieldType’:
dbfopen.c:907:14: note: length computed here
  907 |     if ((int)strlen(pszFieldName) < 10)
      |              ^~~~~~~~~~~~~~~~~~~~

In file included from /usr/include/string.h:495,
                 from dbfopen.c:166:
In function ‘strncpy’,
    inlined from ‘DBFAlterFieldDefn’ at dbfopen.c:2067:9:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: warning: ‘__builtin_strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dbfopen.c: In function ‘DBFAlterFieldDefn’:
dbfopen.c:2066:14: note: length computed here
 2066 |     if ((int)strlen(pszFieldName) < 10)
      |              ^~~~~~~~~~~~~~~~~~~~
```

Curious whether this was addressed upstreams and if so how, looking at the code I realised it may be worth to update it all. After all it is six years since last upgrade.

**Comment by nilason** [comment]
@metzm Perhaps you could take a look at this, that I didn’t overlooked something?

File: lib/external/shapelib/dbfopen.c [hunk 1]
```diff
@@ -34,7 +34,7 @@
  * DEALINGS IN THE SOFTWARE.
  ******************************************************************************/
 
-#include "shapefil.h"
+#include <grass/shapefil.h>
```
Comment (metzm) [inline lib/external/shapelib/dbfopen.c:37]: `#include "shapefil.h"` should be fine because this header is right here?

Comment (nilason) [reply lib/external/shapelib/dbfopen.c]: You're right, I just mechanically reapplied the old patch in this regard. Now, the patches needed is really at minimum.

Updated accordingly.

**Review by metzm — APPROVED** [review]
Please merge!

**Comment by nilason** [comment]
> Please merge!

Done. Thanks!

**Pull request merged**